File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[]}
     */
    productExceptSelf(nums) {
        const res = []
        for (let i = 0; i < nums.length; i++) {
            let localRes
            for (let j = 0; j < nums.length; j++ ) {
                if (i == j) {
                    continue
                }
                if (localRes === undefined) {
                    localRes = nums[j]
                } else {    
                    localRes *= nums[j]
                }
            }
            res.push(localRes)
        }
        return res
    }
}

File: Data Structures & Algorithms/subsets/submission-0.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = []

        subset = []
        # max depth = log2(2^len(nums))=len(nums)
        def backtrack(depth):
            if depth >= len(nums):
                res.append(subset[:])
                return
            
            # branch where we decide to add a number to a subset
            subset.append(nums[depth])
            backtrack(depth + 1)

            subset.pop()
            backtrack(depth + 1)


        backtrack(0)
        return res